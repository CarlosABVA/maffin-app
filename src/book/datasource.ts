import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
