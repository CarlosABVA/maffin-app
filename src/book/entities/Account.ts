import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Entity, JoinColumn,
  ManyToOne,
  Tree,
  TreeParent,
  TreeChildren,
  OneToMany, PrimaryColumn,
} from 'typeorm';

import type Commodity from './Commodity';
import Money from '../Money';
import type Split from './Split';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static TYPES = ['ROOT', 'ASSET', 'BANK', 'CASH', 'EQUITY', 'LIABILITY', 'INCOME', 'EXPENSE', 'MUTUAL', 'STOCK'];

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.MaxLength(2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @v.IsIn(Account.TYPES)
    type!: string;

  @Column({
    type: 'text',
    select: false,
    default: '',
  })
    path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @TreeChildren()
    children!: Account[];

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'commodity_guid' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  /**
   * Note this only works if the entity has access to splits
   */
  get total(): Money {
    if (this.type === 'ROOT') {
      // @ts-ignore
      return null;
    }

    const total = this.splits.reduce(
      (acc, split) => acc.add(
        new Money(split.quantityNum / split.quantityDenom, this.commodity.mnemonic),
      ),
      new Money(0, this.commodity.mnemonic),
    );

    if (total.isNegative()) {
      return total.multiply(-1);
    }

    return total;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });
