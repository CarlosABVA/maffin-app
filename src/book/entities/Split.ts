import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Relation,
  Entity, JoinColumn, ManyToOne, PrimaryColumn,
} from 'typeorm';

import type Account from './Account';
import type Transaction from './Transaction';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE splits (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   tx_guid         CHAR(32) NOT NULL,
 *   account_guid    CHAR(32) NOT NULL,
 *   memo            text(2048) NOT NULL,
 *   action          text(2048) NOT NULL,
 *   reconcile_state text(1) NOT NULL,
 *   reconcile_date  timestamp NOT NULL,
 *   value_num       integer NOT NULL,
 *   value_denom     integer NOT NULL,
 *   quantity_num    integer NOT NULL,
 *   quantity_denom  integer NOT NULL,
 *   lot_guid        CHAR(32)
 * );
*/
@Entity('splits')
export default class Split extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @ManyToOne('Transaction')
  @JoinColumn({ name: 'tx_guid' })
    fk_transaction!: Relation<Transaction> | string;

  get transaction(): Transaction {
    return this.fk_transaction as Transaction;
  }

  @ManyToOne('Account')
  @JoinColumn({ name: 'account_guid' })
    fk_account!: Relation<Account> | string;

  get account(): Account {
    return this.fk_account as Account;
  }

  // For stock operations, action is set to Buy
  @PrimaryColumn({
    type: 'varchar',
    length: 2048,
  })
    action?: string = '';

  @Column({
    type: 'integer',
    name: 'value_num',
  })
    valueNum!: number;

  @Column({
    type: 'integer',
    name: 'value_denom',
  })
    valueDenom!: number;

  @Column({
    type: 'integer',
    name: 'quantity_num',
  })
    quantityNum!: number;

  @Column({
    type: 'integer',
    name: 'quantity_denom',
  })
    quantityDenom!: number;

  /**
   * Returns the value of the transaction in the transaction's currency
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR. This means you spent 1000 EUR to buy
   * 122.8501 of whatever the split's account currency is.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - value in transaction's currency
   */
  get value(): number {
    return this.valueNum / this.valueDenom;
  }

  /**
   * Returns the quantity spent in the currency of the split's account.
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR and the symbol of the account
   * is GOOGL. This means that you spent 1000 EUR to purchase 122.8501 GOOGL
   * stocks.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - quantity spent in purchasing in the split account's currency
   */
  get quantity(): number {
    return this.quantityNum / this.quantityDenom;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Split, 'name', { value: 'Split' });
