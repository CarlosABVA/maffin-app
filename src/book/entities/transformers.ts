import { DateTime } from 'luxon';
import { ValueTransformer } from 'typeorm';

export class DateTimeTransformer implements ValueTransformer {
  // eslint-disable-next-line class-methods-use-this
  from(value: string | null): DateTime | null {
    // For some reason, when using TreeRepository this is null at some point
    if (!value) {
      return null;
    }
    const result = DateTime.fromSQL(value);
    if (!result.isValid) {
      throw new Error(`invalid date: ${result.invalidExplanation}`);
    }
    return result;
  }

  // eslint-disable-next-line class-methods-use-this
  to(value: DateTime): string {
    const result = value.toSQLDate();
    if (result === null) {
      throw new Error(`invalid date: ${value.invalidExplanation}`);
    }
    return result;
  }
}
