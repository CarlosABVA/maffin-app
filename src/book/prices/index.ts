export { default as PriceDB } from './PriceDB';
export { default as PriceDBMap } from './PriceDBMap';
