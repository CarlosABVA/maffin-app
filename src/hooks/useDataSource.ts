import React from 'react';
import { DataSource } from 'typeorm';

import useBookStorage from '@/hooks/useBookStorage';
import { initDB } from '@/book/datasource';

// Not sure if this is the best way to cache the storage...
// need to research
let DATASOURCE: DataSource | null = null;

export default function useDataSource(): [DataSource | null] {
  const [bookStorage] = useBookStorage();
  const [, setDataSource] = React.useState<DataSource | null>(DATASOURCE);
  const isInitializing = React.useRef(false);

  React.useEffect(() => {
    async function load() {
      if (bookStorage && DATASOURCE == null && !isInitializing.current) {
        isInitializing.current = true;
        let start;
        let end;

        start = performance.now();
        const rawBook = await bookStorage.get();
        end = performance.now();

        start = performance.now();
        // We re-check here to avoid race conditions because of
        // downloading book data takes time. Without this
        // we have some reference errors where a component may
        // use one instance another the difference instance which
        // causes synchronization issues.
        if (DATASOURCE === null) {
          DATASOURCE = await initDB(rawBook);
          end = performance.now();
          console.log(`init dataSource: ${end - start}ms`);
        }
        setDataSource(DATASOURCE);
        isInitializing.current = false;
      }
    }

    load();
  }, [bookStorage]);

  return [DATASOURCE];
}
