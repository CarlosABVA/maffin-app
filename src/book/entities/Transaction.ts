import { DateTime } from 'luxon';
import {
  BaseEntity,
  Column, CreateDateColumn, Entity, JoinColumn, ManyToOne, OneToMany,
  PrimaryColumn,
} from 'typeorm';
import * as v from 'class-validator';

import type Commodity from './Commodity';
import type Split from './Split';
import { DateTimeTransformer } from './transformers';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE transactions (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   currency_guid   CHAR(32) NOT NULL,
 *   num             text(2048) NOT NULL,
 *   post_date       timestamp NOT NULL,
 *   enter_date      timestamp NOT NULL,
 *   description     text(2048)
 * );
 */

@Entity('transactions')
export default class Transaction extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
    guid!: string;

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'currency_guid' })
    fk_currency!: Commodity | string;

  get currency(): Commodity {
    return this.fk_currency as Commodity;
  }

  @Column({
    type: 'date',
    transformer: new DateTimeTransformer(),
    name: 'post_date',
  })
    date!: DateTime;

  @CreateDateColumn({
    name: 'enter_date',
  })
    enterDate?: Date;

  @OneToMany('Split', (split: Split) => split.fk_transaction)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    default: '',
  })
  @v.MaxLength(2048)
    description!: string;
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Transaction, 'name', { value: 'Transaction' });
