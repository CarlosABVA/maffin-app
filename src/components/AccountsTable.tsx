import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import { DateTime } from 'luxon';
import Link from 'next/link';

import {
  Account,
  Book,
} from '@/book/entities';
import Money from '@/book/Money';
import { PriceDB } from '@/book/prices';
import type { PriceDBMap } from '@/book/prices';
import Table from '@/components/Table';
import useDataSource from '@/hooks/useDataSource';

export type Record = {
  guid: string,
  name: string,
  type: string,
  total: Money,
  subRows: Record[],
};

export default function AccountsTable(): JSX.Element {
  const [dataSource] = useDataSource();
  const [isLoading, setIsLoading] = React.useState(true);
  const [accountsRows, setAccountsRows] = React.useState<Record[]>([]);

  React.useEffect(() => {
    async function load() {
      if (dataSource) {
        let start;
        let end;
        const books = await Book.find();
        const rootAccount = books[0].root;

        start = performance.now();
        const [rootTree, todayQuotes] = await Promise.all([
          dataSource.manager.getTreeRepository(Account).findDescendantsTree(
            rootAccount,
            {
              relations: ['fk_commodity', 'splits', 'splits.fk_transaction'],
            },
          ),
          PriceDB.getTodayQuotes(),
        ]);
        end = performance.now();
        console.log(`tree render time + today quotes: ${end - start}ms`);

        start = performance.now();
        const rootRecord = buildNestedRows(rootTree, todayQuotes);
        end = performance.now();
        setAccountsRows(rootRecord.subRows);
        setIsLoading(false);
      }
    }

    load();
  }, [dataSource]);

  if (isLoading) {
    return (
      <span>
        Loading...
      </span>
    );
  }

  if (!accountsRows.length) {
    return (
      <span>
        Add accounts to start seeing some data!
      </span>
    );
  }

  return (
    <Table<Record>
      columns={columns}
      data={accountsRows}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
    />
  );
}

const columns: ColumnDef<Record>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    header: '',
    enableSorting: false,
    cell: ({ row }) => (
      <Link href={`/dashboard/accounts/${row.original.guid}`}>
        {row.original.name}
      </Link>
    ),
  },
  {
    accessorFn: (row: Record) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];

/**
 * This function recursively goes through all the accounts so it can build the tree
 * of accounts containing total and subrows so they can be rendered in react table.
 *
 * Note that it interacts with PriceDB so prices can be aggregated to the parent's
 * total in order to display total amounts of assets, income, etc.
 */
function buildNestedRows(current: Account, todayQuotes: PriceDBMap): Record {
  const { children } = current;
  let { total } = current;
  let subRows: Record[] = [];

  if (current.type === 'ROOT') {
    subRows = children.map(child => {
      const childRecord = buildNestedRows(child, todayQuotes);
      return childRecord;
    });
  } else {
    const commodity = current.commodity.mnemonic;
    if (['STOCK', 'MUTUAL'].includes(current.type)) {
      const price = todayQuotes.getStockPrice(
        current.commodity.mnemonic,
        DateTime.now(),
      );
      total = total.convert(price.currency.mnemonic, price.value);
    }

    subRows = children.map(child => {
      const childRecord = buildNestedRows(child, todayQuotes);
      let childTotal = childRecord.total;
      if (childTotal.currency !== commodity) {
        const price = todayQuotes.getPrice(
          childTotal.currency,
          commodity,
          DateTime.now(),
        );
        childTotal = childTotal.convert(commodity, price.value);
      }
      total = total.add(childTotal);
      return childRecord;
    });
  }

  return {
    guid: current.guid,
    name: current.name,
    type: current.type,
    total,
    subRows,
  };
}
