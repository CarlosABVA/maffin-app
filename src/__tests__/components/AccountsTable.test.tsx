import React from 'react';
import { render, screen } from '@testing-library/react';
import type { DataSource, BaseEntity } from 'typeorm';

import AccountsTable, { Record } from '@/components/AccountsTable';
import type { TableProps } from '@/components/Table';
import { PriceDB, PriceDBMap } from '@/book/prices';
import { Book, Account } from '@/book/entities';
import Money from '@/book/Money';
import * as dataSourceHooks from '@/hooks/useDataSource';

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('@/components/Table', () => {
  function Table({ columns, data, initialSort }: TableProps<Record>) {
    return (
      <div data-testid="AccountsTable" className="Table">
        <span>{JSON.stringify(columns)}</span>
        <span data-testid="data">{JSON.stringify(data)}</span>
        <span>{JSON.stringify(initialSort)}</span>
      </div>
    );
  }

  return Table;
});

describe('AccountsTable', () => {
  let mockGetTodayQuotes: jest.SpyInstance;

  beforeEach(() => {
    mockGetTodayQuotes = jest.spyOn(PriceDB, 'getTodayQuotes').mockResolvedValue({
      getPrice: (from, to, date) => ({
        guid: 'price_guid',
        value: 0.98,
      }),
    } as PriceDBMap);
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([null]);
    jest.spyOn(Book, 'find').mockResolvedValue([
      {
        // @ts-ignore
        root: {
          name: 'Root',
          currency: Promise.resolve(null),
          total: null,
          type: 'ROOT',
        } as Account,
      },
    ]);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders empty when no dataSource', () => {
    const { container } = render(<AccountsTable />);

    expect(container).toMatchSnapshot();
  });

  it('renders message when no accounts', async () => {
    // @ts-ignore
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{
      manager: {
        getTreeRepository: (e: BaseEntity) => ({
          findDescendantsTree: async () => ({
            name: 'Root',
            currency: null,
            total: null,
            type: 'ROOT',
            children: [],
          }),
        }),
      },
    } as DataSource]);
    render(<AccountsTable />);

    await screen.findByText('Add accounts to start seeing some data!');
  });

  it('builds subRows as expected with non investment accounts', async () => {
    // @ts-ignore
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{
      manager: {
        getTreeRepository: (e: BaseEntity) => ({
          findDescendantsTree: async () => ({
            name: 'Root',
            currency: null,
            total: null,
            type: 'ROOT',
            children: [
              {
                name: 'Assets',
                total: new Money(10, 'EUR'),
                currency: {
                  mnemonic: 'EUR',
                },
                commodity: {
                  mnemonic: 'EUR',
                },
                type: 'ASSET',
                children: [
                  {
                    name: 'Bank',
                    total: new Money(1000, 'USD'),
                    currency: {
                      mnemonic: 'USD',
                    },
                    commodity: {
                      mnemonic: 'USD',
                    },
                    type: 'BANK',
                    children: [],
                  },
                ],
              },
            ],
          }),
        }),
      },
    } as DataSource]);

    const { container } = render(<AccountsTable />);
    await screen.findByTestId('data');

    expect(container).toMatchSnapshot();
    expect(mockGetTodayQuotes).toBeCalledTimes(1);
  });

  it('builds subRows as expected with investment accounts', async () => {
    mockGetTodayQuotes = jest.spyOn(PriceDB, 'getTodayQuotes').mockResolvedValue({
      getPrice: (from, to, date) => ({
        guid: 'price_guid',
        value: 0.98,
      }),
      getStockPrice: (from, date) => ({
        guid: 'stock_price_guid',
        value: 100,
        currency: {
          mnemonic: 'USD',
        },
      }),
    } as PriceDBMap);
    // @ts-ignore
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{
      manager: {
        getTreeRepository: (e: BaseEntity) => ({
          findDescendantsTree: async () => ({
            name: 'Root',
            currency: null,
            total: null,
            type: 'ROOT',
            children: [
              {
                name: 'Stocks',
                total: new Money(0, 'EUR'),
                currency: {
                  mnemonic: 'EUR',
                },
                commodity: {
                  mnemonic: 'EUR',
                },
                type: 'ASSET',
                children: [
                  {
                    name: 'GOOGL',
                    total: new Money(2, 'GOOGL'),
                    currency: {
                      mnemonic: 'USD',
                    },
                    commodity: {
                      mnemonic: 'GOOGL',
                    },
                    type: 'STOCK',
                    children: [],
                  },
                ],
              },
            ],
          }),
        }),
      },
    } as DataSource]);

    const { container } = render(<AccountsTable />);
    await screen.findByTestId('data');

    expect(container).toMatchSnapshot();
    expect(mockGetTodayQuotes).toBeCalledTimes(1);
  });
});
