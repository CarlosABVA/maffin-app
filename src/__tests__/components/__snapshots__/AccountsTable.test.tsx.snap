// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsTable builds subRows as expected with investment accounts 1`] = `
<div>
  <div
    class="Table"
    data-testid="AccountsTable"
  >
    <span>
      [{"accessorKey":"name","id":"name","header":"","enableSorting":false},{"id":"total","header":""}]
    </span>
    <span
      data-testid="data"
    >
      [{"name":"Stocks","type":"ASSET","total":{"_raw":{"amount":1960000,"currency":{"code":"EUR","base":10,"exponent":2},"scale":4}},"subRows":[{"name":"GOOGL","type":"STOCK","total":{"_raw":{"amount":20000,"currency":{"code":"USD","base":10,"exponent":2},"scale":2}},"subRows":[]}]}]
    </span>
    <span>
      {"id":"total","desc":true}
    </span>
  </div>
</div>
`;

exports[`AccountsTable builds subRows as expected with non investment accounts 1`] = `
<div>
  <div
    class="Table"
    data-testid="AccountsTable"
  >
    <span>
      [{"accessorKey":"name","id":"name","header":"","enableSorting":false},{"id":"total","header":""}]
    </span>
    <span
      data-testid="data"
    >
      [{"name":"Assets","type":"ASSET","total":{"_raw":{"amount":99000,"currency":{"code":"EUR","base":10,"exponent":2},"scale":2}},"subRows":[{"name":"Bank","type":"BANK","total":{"_raw":{"amount":1000,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"subRows":[]}]}]
    </span>
    <span>
      {"id":"total","desc":true}
    </span>
  </div>
</div>
`;

exports[`AccountsTable renders empty when no dataSource 1`] = `
<div>
  <span>
    Loading...
  </span>
</div>
`;
